### DISCUSSION TITLE
Do we have any utility like lowpass, highpass and bandpass filter in xCDAT?

### DISCUSSION THREAD

**Discussion by JiveshDixit** [xCDAT/xcdat#431] [2. Questions]
These filters could be a really useful in many applications in climate analysis. I was wondering as xCDAT being a climate data analysis tool, does it have utility to perform the same? If not, is there any plan to add that in future? Seems we have a xarray based package "xrscipy" that does the same.

**Comment by pochedls** [comment]
Hi @JiveshDixit – I agree that filtering is really useful. Unfortunately, xCDAT doesn't have any filtering capabilities at this time. A good place to discuss the potential for this kind of functionality is this thread: https://github.com/xCDAT/xcdat/discussions/271. In general, xCDAT tries to ensure feature are maintainable and aren't already implemented in another package. Thanks for pointing out xrscipy! @lee1043 - may also have comments (not sure if PMP uses any temporal filtering).

**Comment by chengzhuzhang** [comment]
Thanks for the discussion! NCL has the [filter functions ](https://www.ncl.ucar.edu/Applications/filter.shtml)which has been really helpful. It would be really nice to have a python implementation of these libraries, I'm not sure if there is a tool/package that covers these functions exist...

**Comment by lee1043** [comment]
Thank you for the discussion, @JiveshDixit. I ditto what @pochedls said. Few metrics of PMP use some features like filtering, e.g., in its monsoon and precipitation variability metric (@msahn please correct me if I am wrong). I agree filtering would be useful but not sure if we would want to have it in xcdat if xr-scipy can well handle it (Thank you @JiveshDixit for bringing xr-scipy to our attention).

**Reply by msahn** [reply]
The precipitation variability metric uses a [power spectrum provided by SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.welch.html), but filtering is not used in the metric.

**Comment by mfwehner** [comment]
You could use os.sys to just make a call to the previously mentioned NCL routines through some intermediate files too.
Michael

**Reply by chengzhuzhang** [reply]
Thank you @mfwehner for pointing out this workaround. I can't seem to find a tutorial for calling the NCL routines from Python. Having some examples would be really great.

**Comment by mfwehner** [comment]
What I would do is write out a netcdf file in the python first.
Then
import os
os.system(’ncl ‘+ncl_script_name+’ ‘+arguments)

Where ncl_script_name is your script and arguments are whatever input is required).
Then reopen the ncl output netcdf file and reading back in the  results you need.

A bit clumsy because you are transferring information indirectly via files, but it works :)

Michael

**Reply by chengzhuzhang** [reply]
Thank you @mfwehner!
If we were going to provide the filters (i.g., Lanczos and Butterworth filters, as those supported by NCL) through xcdat, what you proposed is a good way to test and compare with NCL results. 
Personally, I think the implementation won't be too bad, but could be popular. Alternatively, it might make more sense to contribute to xr-scipy with filters that are important (in climate analysis) but not already included (admittedly I need to learn more about those different filter types..).

**Comment by mfwehner** [comment]
Jill
One argument for using ncl routines in general, is that there is a large climate & weather analyst user base and this library is well vetted. I don’t know about xr-scipy. It may be larger. Or not, if it is newer.
Michael

**Comment by tomvothecoder** [comment]
Just an FYI, [geocat-comp](https://github.com/NCAR/geocat-comp) (a similar analysis package) has these filtering features: https://geocat-comp.readthedocs.io/en/latest/user_api/index.html#fourier-filters

**Comment by jypeter** [comment]
One of our interns had a similar problem. He wanted to run a highpass filter (based on some lanczos stuff. He can be more specific when he find his GH identifier), on a long time series split in several nc files

He had a working solution, doing things the hard way (looping over the files by hand, etc...), and I told him to at least experiment with `xc.open_mfdataset`. This seemed to work and not use too much memory, but applying the filter on this virtually aggregated dataset was slower than doing it on data concatenated in memory

He his going to look at the `geocat.comp` based solution suggested by @tomvothecoder just above (and I have told him to also look at the [xrft](https://xrft.readthedocs.io/en/latest/index.html) package), but maybe somebody has come up with something more recent for "cleanly and efficiently applying a filter to a long time series of 2D data"

**Reply by pochedls** [reply]
@jypeter – possibly a longshot, but sometimes I find that calling `.load()` on the dataset (or subsetted dataset) to put it into memory results in faster computations compared to letting xarray decide when to load data from the virtual object. This might be worth a try as a simple solution (assuming you can fit the dataset into memory).

**Comment by jypeter** [comment]
Thanks @pochedls , this `load` stuff is good to know !

But it would be nice to keep a virtual object to reduce memory consumption **and** still get good I/O and computing performance. Is there a way to tell xarray to buffer/preload (only) part of the data ?

Of course xarray is not aware of the kind of operation we are going to perform on the data, and how much ram we can use on the server. But in the case (running) averages or filters where we know that we are not going to use time steps separated by more than N steps, there might be a way to give xarray a hint

Actually, if we have aggregated NN files, and that each file has MM time steps, and we need LL consecutive time steps for the computation, we could tell the aggregation routine that we need loaded data from no more than FF consecutive open files. I'm not sure I'm making sense because it's very hot here, with no air conditioning

There is no rush for a solution, but this seems to be a common use case for our type of (climate model) data. Hopefully our intern will stay 3 more years, to get a PhD, and can experiment with this

**Reply by pochedls** [reply]
I am not sure how to only pre-load a select part of the dataset (you could use `.sel` or `.isel`, but that would drop the rest of the data). 

One other approach might be to re-[`.chunk`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.chunk.html) your data. xarray is supposed to load/operate on these chunks in parallel. This works out-of-the-box with some functions, but other use-cases require more work to get working (`apply_ufunc` and/or `map_blocks`). There is some documentation about this [here](https://docs.xarray.dev/en/stable/user-guide/dask.html). I haven't used these features (I tend to explicitly load/process chunks and re-concatenate at the end), but they might work for you. 

@tomvothecoder has more experience with this and can probably add more context or point to better examples.
